app/models/segmentation_statistic.rb

```diff
@@ -0,0 +1,602 @@
+# frozen_string_literal: true
+
+class SegmentationStatistic < ApplicationRecord
+  def self.generate_stats_for_day( at_time = Time.now, options = {} )
+    at_time = at_time.utc.end_of_day
+    if options[:force]
+      SegmentationStatistic.where( "DATE(created_at) = DATE(?)", at_time.utc ).delete_all
+    elsif stats_generated_for_day?( at_time )
+      return
+    end
+    sleep 1
+    SegmentationStatistic.create!(
+      data: generate_monthly_segmentation_metrics( at_time ),
+      created_at: at_time.beginning_of_day
+    )
+  end
+
+  def self.stats_generated_for_day?( at_time = Time.now )
+    SegmentationStatistic.where( "DATE(created_at) = DATE(?)", at_time.utc ).exists?
+  end
+
+  def self.get_viz_stats_for_day( at_time = Time.now )
+    stats = SegmentationStatistic.where( "DATE(created_at) = DATE(?)", at_time.utc )
+    format_for_viz( stats.first.data ) if stats.exists?
+  end
+
+  # private
+
+  # Generate monthly metrics
+  def self.generate_monthly_segmentation_metrics( at_time = Time.now )
+    monthly_segmentation_data = generate_monthly_segmentation_data( at_time )
+    generate_main_metrics( monthly_segmentation_data )
+  end
+
+  # Generate all the users data for a month
+  def self.generate_monthly_segmentation_data( at_time = Time.now )
+    monthly_segmentation_data = {}
+    day = at_time
+    30.times do
```

Comment:
To me this method is misnamed, or its not looking at the right time frame. This is generating data for the last 30 days from the given day, it does not generate data for a calendar month. It may be getting data from multiple months or part of a month